library/core/src/needle/needle.rs

```diff
@@ -0,0 +1,398 @@
+use super::haystack::{Hay, Haystack, Span};
+
+use crate::ops::Range;
+
+/// A searcher, for searching a [`Needle`] from a [`Hay`].
+///
+/// This trait provides methods for searching for non-overlapping matches of a
+/// needle starting from the front (left) of a hay.
+///
+/// # Safety
+///
+/// This trait is marked unsafe because the range returned by its methods are
+/// required to lie on valid codeword boundaries in the haystack. This enables
+/// users of this trait to slice the haystack without additional runtime checks.
+///
+/// # Examples
+///
+/// Implement a searcher and consumer which matches `b"Aaaa"` from a byte string.
```

Comment:
Maybe we should explain what is a consumer? I don't know if consumer can be an easy term to understand since there are many other ways that a consumer can be used, I wonder if there is a better term that can be used here.

Reply:
I think i'll just link searcher and consumer to the corresponding trait...